### PR TITLE
Add nbconvert preprocessor

### PR DESCRIPTION
**Pull request by davidbrochart**

### CODE REVIEW

**Comment by davidbrochart** [comment]
The tests fail on Windows because nbconvert needs pandoc, and I'm not sure how to install that on Windows.

Otherwise, it looks like this:
image

**Comment by SylvainCorlay** [comment]
Why not using conda for CI?

**Comment by davidbrochart** [comment]
Azure Pipelines doesn't let me see the log, so it's hard to debug.

File: tests/notebook.py [hunk 1]
```diff
@@ -1,11 +1,37 @@
...
+
+        c = Config()
+        c.MemestraDeprecationChecker.decorator = ('decoratortest', 'deprecated')
+        c.RSTExporter.preprocessors = [preprocessor.MemestraDeprecationChecker]
+
+        deprecation_checker = RSTExporter(config=c)
+
+        rst = deprecation_checker.from_notebook_node(notebook)[0]
+
+        with open('tests/misc/memestra_nb_demo.rst') as f:
```
Comment (serge-sans-paille) [inline tests/notebook.py:29]: use `os.path.join` for more portability

File: tests/misc/memestra_nb_demo.ipynb [hunk 2]
```diff
@@ -2,18 +2,9 @@
  "cells": [
```
Comment (serge-sans-paille) [inline tests/misc/memestra_nb_demo.ipynb:2]: Why are these changes needed?

File: memestra/nbmemestra.py [hunk 3]
```diff
@@ -4,9 +4,13 @@
 import nbformat
 
 
-def nbmemestra(nbfile, decorator):
+def nbmemestra(nb_or_file, decorator):
+    if isinstance(nb_or_file, nbformat.NotebookNode):
```
Comment (serge-sans-paille) [inline memestra/nbmemestra.py:8]: I'd rather avoid that kind of checks and accept a single kind of node, is it possible to update the call  site instead?

File: memestra/preprocessor.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+            deprecations[i] = deprecation
+        i = 0
+        for cell in nb.cells:
+            if cell['cell_type'] == 'code':
+                if i in deprecations:
+                    for d in deprecations[i]:
+                        line = d[2]
+                        col = d[3]
+                        outputs = cell.get('outputs', [])
+                        message = 'On line {}:\n'.format(line)
```
Comment (serge-sans-paille) [inline memestra/preprocessor.py:27]: This is difficult to read. Maybe a multiline string with a single format call would help?

File: memestra/preprocessor.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+            i = int(d[1].split('[', 1)[1].split(']')[0])
+            deprecation = deprecations.get(i, [])
+            deprecation.append(d)
+            deprecations[i] = deprecation
+        i = 0
+        for cell in nb.cells:
+            if cell['cell_type'] == 'code':
+                if i in deprecations:
+                    for d in deprecations[i]:
+                        line = d[2]
```
Comment (serge-sans-paille) [inline memestra/preprocessor.py:24]: nits: ``line, col = d[2:4]``

File: memestra/preprocessor.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        deprecations = {}
+        for d in nbmemestra.nbmemestra(nb, self.decorator):
+            i = int(d[1].split('[', 1)[1].split(']')[0])
+            deprecation = deprecations.get(i, [])
+            deprecation.append(d)
+            deprecations[i] = deprecation
+        i = 0
+        for cell in nb.cells:
+            if cell['cell_type'] == 'code':
+                if i in deprecations:
```
Comment (serge-sans-paille) [inline memestra/preprocessor.py:22]: *warning* you're reusing the `i` identifier that's already used at line 19 :-/

File: memestra/preprocessor.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+class MemestraDeprecationChecker(Preprocessor):
+    """A memestra-based preprocessor that checks code for places where deprecated functions are called"""
+
+    decorator = Tuple(help="path to import the deprecation decorator, e.g. ('module', 'attribute')").tag(config=True)
+
+    def preprocess(self, nb, resources):
+        self.log.info("Using decorator '%s.%s' to check deprecated functions", self.decorator[0], self.decorator[1])
+        deprecations = {}
+        for d in nbmemestra.nbmemestra(nb, self.decorator):
+            i = int(d[1].split('[', 1)[1].split(']')[0])
```
Comment (serge-sans-paille) [inline memestra/preprocessor.py:15]: you should support decorator with more than 2 items

File: memestra/preprocessor.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+            deprecation = deprecations.get(i, [])
+            deprecation.append(d)
+            deprecations[i] = deprecation
+        i = 0
```
Comment (serge-sans-paille) [inline memestra/preprocessor.py:19]: please rename this in `nb_code_cell` or something explicit like this.

File: memestra/nbmemestra.py (hunk 3 above)
Comment (davidbrochart) [reply memestra/nbmemestra.py:8]: I know, but the preprocessor framework (see https://nbconvert.readthedocs.io/en/latest/nbconvert_library.html#Custom-Preprocessors) forces me to have a `preprocess` method that accepts an object of class `NotebookNode`, in which I call `nbmemestra`. The other option is to have `nbmemestra` accept a file and call a `nbmemestra_from_nbnode` or something like that, what do you think?

File: memestra/preprocessor.py (hunk 7 above)
Comment (davidbrochart) [reply memestra/preprocessor.py:15]: :+1:

File: memestra/preprocessor.py (hunk 8 above)
Comment (davidbrochart) [reply memestra/preprocessor.py:19]: :+1:

File: memestra/preprocessor.py (hunk 5 above)
Comment (davidbrochart) [reply memestra/preprocessor.py:24]: :+1:

File: memestra/preprocessor.py (hunk 4 above)
Comment (davidbrochart) [reply memestra/preprocessor.py:27]: :+1:

File: tests/misc/memestra_nb_demo.ipynb (hunk 2 above)
Comment (davidbrochart) [reply tests/misc/memestra_nb_demo.ipynb:2]: I wanted to have a deprecated function call on a column other than 0, to check that also. I cleared all the outputs but you're right, it can be interesting to check the mix of different outputs, from the deprecation function and the preprocessor.

File: tests/notebook.py (hunk 1 above)
Comment (davidbrochart) [reply tests/notebook.py:29]: :+1:

File: memestra/nbmemestra.py [hunk 9]
```diff
@@ -35,5 +35,10 @@ def nbmemestra(nbfile, decorator):
     return result
 
 
+def nbmemestra(nbfile, decorator):
```
Comment (serge-sans-paille) [inline memestra/nbmemestra.py:38]: Yeah, I prefer that version, thanks.

File: memestra/preprocessor.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        deprecations = {}
+        for d in nbmemestra.nbmemestra_from_nbnode(nb, self.decorator):
+            code_cell_i = int(re.search(r"\[(\d+)\]", d[1]).group(1))
+            deprecation = deprecations.get(code_cell_i, [])
+            deprecation.append(d)
+            deprecations[code_cell_i] = deprecation
+        # filter out non-code cell as it is done in nbmemestra
+        code_cell_i = 0
+        for cell in nb.cells:
+            if cell['cell_type'] == 'code':
```
Comment (serge-sans-paille) [inline memestra/preprocessor.py:23]: nits: To avoid deep nesting, it is common practice to have early exits, you can switch to

```python
if cell['cell_type'] != 'code':
    continue
if code_cell_i not in deprecations:
    continue

for d in deprecations[code_cell_i]:
   ...
```

File: tests/notebook.py [hunk 11]
```diff
@@ -1,11 +1,33 @@
...
+
+        c = Config()
+        c.MemestraDeprecationChecker.decorator = ('decoratortest', 'deprecated')
+        c.RSTExporter.preprocessors = [preprocessor.MemestraDeprecationChecker]
+
+        deprecation_checker = RSTExporter(config=c)
+
+        rst = deprecation_checker.from_notebook_node(notebook)[0]
+
+        with open(os.path.join('tests', 'misc', 'memestra_nb_demo.rst')) as f:
```
Comment (serge-sans-paille) [inline tests/notebook.py:30]: This makes a few assumption on `cwd`. You should probably take the `os.basedir` of `__path__` and make your test relative to that. (same for all the paths you add btw)

File: memestra/preprocessor.py (hunk 10 above)
Comment (davidbrochart) [reply memestra/preprocessor.py:23]: :+1:

File: tests/notebook.py (hunk 11 above)
Comment (davidbrochart) [reply tests/notebook.py]: :+1:

**Comment by serge-sans-paille** [comment]
the azure test all fail but travis is OK... @JohanMabille ?

**Comment by davidbrochart** [comment]
The test now ignores the differences in the rst file that only consist of empty lines, since it looks like nbconvert behaves differently on windows and on linux in that regard.
